### PR TITLE
Proc macro multi abi proof of concept

### PR DESCRIPTION
**Pull request by alexjg**
#8925 was irritating me so I thought I would have a bash at fixing it. What I've done here is copy the `crates/proc_macro_srv/src/proc_macro` code (which is copied from `<RUST>/library/proc_macro`) to `crates/proc_macro_srv/src/proc_macro_nightly` and the modified the nightly version to include the changes from https://github.com/rust-analyzer/rust-analyzer/pull/9047 and https://github.com/yotamofek/rust-analyzer/commit/aeb7b183a2302af968f1110fb3cbd00c76a34974

This gives us the code to support both stable and nightly ABIs. Then we use the `proc_macro_api::version::read_dylib_info` to determine which version of the ABI to load when creating a `ProcMacroLibraryLibLoading` (which is now an enum). 

This seems to work for me.  The code could be cleaned up but I wanted to see if the approach makes sense before I spend more time on it.

I've split the change into two commits, the first is just copying and modifying the `proc_macro` crate, the second contains most of the interesting work around figuring out which ABI to use.

### CODE REVIEW

File: crates/proc_macro_srv/src/dylib.rs [hunk 1]
```diff
@@ -74,26 +83,117 @@ fn load_library(file: &Path) -> Result<Library, libloading::Error> {
...
+enum ProcMacroABI {
+    Stable,
+    Nightly,
+}
+
+impl TryFrom<RustCInfo> for ProcMacroABI {
+    type Error = LoadProcMacroDylibError;
+
+    fn try_from(info: RustCInfo) -> Result<Self, Self::Error> {
+        if info.version.0 < 1 {
```
Comment (bjorn3) [inline crates/proc_macro_srv/src/dylib.rs:95]:
```suggestion
        if info.version.0 != 1 {
```

**Comment by matklad** [comment]
I think it all makes sense, yeah! 

One thing I worry is that, by copy-pasting from rustc more, it will be hard to maintain this in the long run, as few people understand how this thing works process wise. 

I suggest merging this on Monday (after the release, to get more testing time), and then, if anyone wants to:

* automate the copy-pasting by adding `cargo xtask update-syntax-bridge` command  or an ignored test a-la [sourcegen_int_completion](https://github.com/rust-analyzer/rust-analyzer/blob/2e52d6ea938129e67b567e1310cf43976c036786/crates/ide_completion/src/tests/sourcegen.rs#L8-L13)
* refeactor the rustc side of things, such that the syntax bridge is auto-published and is compatible with stable.

**Comment by bjorn3** [comment]
proc_macro should be autopublished as rustc-ap-proc_macro I think. It doesn't compile on stable though as it uses a lot of unstable features: https://github.com/rust-lang/rust/blob/master/library/proc_macro/src/lib.rs#L20-L33

**Comment by matklad** [comment]
Yeah, we def don't want to de-facto stabilize the whole of proc-macro. I guess, the pre-requisite here is extracting the `bridge` module to a stand-alone crate first.

**Comment by alexjg** [comment]
One thing that I am concerned about with this approach is that the ABI selection is done by this piece of code:

```
if info.version.0 < 1 {
    Err(LoadProcMacroDylibError::UnsupportedABI)
} else if info.version.1 < 47 {
    Err(LoadProcMacroDylibError::UnsupportedABI)
} else if info.version.1 < 54 {
    Ok(ProcMacroABI::Stable)
} else {
    Ok(ProcMacroABI::Nightly)
}
```

I think this means that every time there is a version bump in the stable rust compiler we will need to increment the version in the third branch of this expression right?

**Comment by flodiebold** [comment]
I think the problem is more the naming of the ABIs; what's currently the "nightly" ABI will become the "stable" one at some point. If we want to support the latest stable and nightly all the time, that's probably the best approach though; when the stable with the changed ABI is released, we'll remove "stable", copy "nightly" to "stable" and make `< 54` unsupported as well. Or we could keep supporting older stable versions, but I'm not sure we want to do that.

**Comment by lnicola** [comment]
Maybe we can name them by their release version?

**Comment by flodiebold** [comment]
You mean `1_47` and `1_54`? That would make sense as well, I think.

**Comment by lnicola** [comment]
Yeah, and update `1_54` during the release cycle if it changes more than once.

**Comment by flodiebold** [comment]
Yeah, that might be the best variant. And if there's a new ABI break in the future, we can introduce e.g. `1_56`, and then separately think about whether we want to limit the number of ABIs we support to two or keep all of them.

**Comment by alexjg** [comment]
I like it, alright, let me just whip something up.

**Comment by alexjg** [comment]
I've reorganised things a bit so that it should be easier to add ABIs in the future. Take a look.

**Comment by Dessix** [comment]
> Yeah, and update `1_54` during the release cycle if it changes more than once.

If an ABI changes more than once in nightly (Has this ever happened before?), would it be better to make the tuple a triple, and have the third member be either the variant or the date-of-change?

Many repositories- my own included- pin a working nightly in `rust-toolchain` to avoid breakages in CI. If an ABI changes multiple times in the same target version, we could still end up with non-functioning `rust-toolchain` options despite having gone through the effort of making various ABIs work.

As for how many versions to support- what about limiting ourselves to two stabilized ABI versions and "all(ish) nightly ABI versions that haven't reached stable"?

**Comment by lnicola** [comment]
I think it's fine to support only one nightly (preferably the latest).

**Comment by Dessix** [comment]
> Did you mean the nightly versions in the current release cycle, or all the nightly versions ever?
I meant all nightly ABI versions that haven't reached stable yet. As for the "ish", I mean that any that are deemed too broken or too short-lived could be skipped.

**Comment by lnicola** [comment]
> I meant all nightly ABI versions that haven't reached stable yet.

In the current cycle (1.55) or also the nightlies from the previous release cycles?

**Comment by Dessix** [comment]
I was advocating for the current release cycle. I wasn't aware old release cycles were a thing people used, but I've never used stable for longer than the first few `rustup` commands. If a nightly ABI variant is never going to reach stable, there's little sense in maintaining support for it.

**Comment by lnicola** [comment]
> Many repositories- my own included- pin a working nightly in rust-toolchain to avoid breakages in CI.

Which nightly are you using now? Pretty much every project I've seen doing this was using an older nightly.

> If a nightly ABI variant is never going to reach stable, there's little sense in maintaining support for it once stable passes it up.

Yeah, but a nightly ABI is even more temporary than a stable one (stable is forever, nightlies are only relevant for a couple of days in the model above). A workaround would be to use `rust-analyzer-preview` from that specific nightly, which is still not great, but presumably temporary.

**Comment by Dessix** [comment]
> Which nightly are you using now? Pretty much every project I've seen doing this was using an older nightly.

2021-05-17, the newest I found that worked with Rust-Analyzer still supporting proc-macros. Normally, we'd adjust the pin forward every other week or two.

Supporting every nightly is one thing- but the ABI from 5-17 remains functional with rust-analyzer builds from 2 months later- so I suspect they either don't change often enough to cause problems, or something else is going on that I don't understand, but it doesn't sound - to me - like the cutoff date needs to be anywhere in the realm of "days".

**Comment by lnicola** [comment]
So the two changes in the current cycle were merged on May 20 and 30. Assuming we can summon alexjg to update this right after a breaking change is merged, I'm not sure it's worth supporting those nightlies now in July.

**Comment by Dessix** [comment]
Fair- I'm not suggesting we backdate, but with only a few breaks every few months, it seems like it should be viable to support future changes without immediately dumping the older variants, since the code will already be written if they were in any way long-lived.

**Comment by lnicola** [comment]
> The biggest downside of not dropping ABIs as we go forward is probably compilation times no?

Yes, and some interference with the CI metrics (memory, build time, type inference statistics). Of course, those aren't perfect anyway, and they change from one `rustc` release to another.

> with only a few breaks every few months, it seems like it should be viable to support future changes without immediately dumping the older variants

It's a matter of policy: we already only support the latest stable compiler and last two VS Code releases.

It's one thing to say "we support the latest stable, beta, and some recent nightly version" and something else to say "we support stable, beta, and every nightly in the last three months". What if we don't get to implement one ABI version before a second breaking change gets merged?

**Comment by matklad** [comment]
I think a reasonable stance is to initially support only the latest nightly, and see how it goes. If there's a huge demand for supporting more than one nightly, and volontaires to do the supporting work, we can add more! But starting small let's us to better estimate the maintenance/benefit tradeoff. 

bors r+

**Comment by flodiebold** [comment]
If we support more ABIs, it also increases the effort when we need to change non-ABI-related duplicated code, like the `expand` and `list_macros` functions. Maybe we can reduce that duplication though.

**Pull request merged**